cpp/ql/test/query-tests/Likely Bugs/Memory Management/UsingExpiredStackAddress/test.cpp

```diff
@@ -0,0 +1,183 @@
+struct S100 {
+  int i;
+  int* p;
+};
+
+static struct S100 s101;
+
+void escape1() {
+  int x;
+  s101.p = &x;
+}
+
+int simple_field_bad() {
+  escape1();
+  return *s101.p; // BAD
+}
+
+int simple_field_good() {
+  escape1();
+  return s101.i; // GOOD
+}
+
+int deref_p() {
+  return *s101.p;
+}
+
+int field_indirect_bad() {
+  escape1();
+  return deref_p(); // BAD [NOT DETECTED]
+}
+
+int deref_i() {
+  return s101.i;
+}
+
+int field_indirect_good() {
+  escape1();
+  return deref_i(); // GOOD
+}
+
+void store_argument(int *p) {
+  s101.p = p;
+}
+
+int store_argument_value() {
+  int x;
+  store_argument(&x);
+  return *s101.p; // GOOD
+}
+
+void store_address_of_argument(int x) {
+  s101.p = &x;
+}
+
+int store_argument_address() {
+  int x;
+  store_address_of_argument(x);
+  return *s101.p; // BAD
+}
+
+void address_escapes_through_pointer_arith() {
+  int x = 0;
+  int* p0 = &x;
+  int* p1 = p0 + 1;
+  int* p2 = p1 - 1;
+  int* p3 = 1 + p2;
+  p3++;
+  s101.p = p3;
+}
+
+int test_pointer_arith_bad() {
+  address_escapes_through_pointer_arith();
+  return *s101.p; // BAD
+}
+
+int test_pointer_arith_good_1() {
+  int x;
+  address_escapes_through_pointer_arith();
+  s101.p = &x;
+  return *s101.p; // GOOD [FALSE POSITIVE]
+}
+
+int test_pointer_arith_good_2(bool b) {
+  int x;
+  if(b) {
+    address_escapes_through_pointer_arith();
+  }
+  return *s101.p; // GOOD (we can't say for sure that this is a local address)
+}
+
+void field_address_escapes() {
+  S100 s;
+  s101.p = &s.i;
+}
+
+int test_field_address_escapes() {
+  field_address_escapes();
+  return s101.p[0]; // BAD
+}
+
+void escape_through_reference() {
+  int x = 0;
+  int& r0 = x;
+  int& r1 = r0;
+  r1++;
+  s101.p = &r1;
+}
+
+int test_escapes_through_reference() {
+  escape_through_reference();
+  return *s101.p; // BAD
+}
+
+struct S300 {
+  int a1[15];
+  int a2[14][15];
+  int a3[13][14][15];
```

Comment:
These do not seem to be used below?

Reply:
Oops. I've added those tests in 53299d61eb2b7fd61a705da6b3703ec07c622842 now.